### PR TITLE
Fix chroot detection

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

**Review by bluca — APPROVED** [review]

File: src/basic/virt.c [hunk 1]
```diff
@@ -867,10 +867,16 @@ int running_in_userns(void) {
...
 
+        /* If we're PID1, /proc may not be mounted (and most likely we're not in a chroot). But PID1 will
+         * mount /proc, so all other programs can assume that if /proc is *not* available, we're in some
+         * chroot. */
+
         if (getenv_bool("SYSTEMD_IGNORE_CHROOT") > 0)
                 return 0;
 
         r = files_same("/proc/1/root", "/", 0);
+        if (r == -ENOENT && getpid_cached() != 1)
```
Comment (poettering) [inline src/basic/virt.c]: i find this quite a stretch, but maybe this is ok.

but please tighten the check, i.e. validate the -ENOENT with proc_mounted(), and output a debug message if we determine this case to be triggered explaining things.

**Comment by keszybz** [comment]
Updated with the requested changed.

I dropped the cleanup commit from this PR, I'll submit it with some other typo fixes and such.

**Comment by mrc0mmand** [comment]
(Ignore the TEST-73-LOCALE fails in Arch jobs, they're not related to this PR)

File: src/basic/virt.c [hunk 2]
```diff
@@ -867,10 +867,23 @@ int running_in_userns(void) {
 int running_in_chroot(void) {
         int r;
 
+        /* If we're PID1, /proc may not be mounted (and most likely we're not in a chroot). But PID1 will
+         * mount /proc, so all other programs can assume that if /proc is *not* available, we're in some
+         * chroot. */
```
Comment (poettering) [inline src/basic/virt.c:872]: actually, if we are PID 1 then the comparison of `/proc/1/root` with `/` doesn't really make much sense, it's of course going to be true, we compare our root with our root, wich is definitely true ;-)

Hence, maybe add a check early on:

```c
if (getpid_cached() == 1)
        return false;
```

and then drop the getpid_cached() below

Comment (keszybz) [reply src/basic/virt.c:872]: Updated.

**Comment by keszybz** [comment]
Please merge.

**Review by bluca — APPROVED** [review]

**Pull request merged**